extensions/amp-instagram/1.0/component.js

```diff
@@ -15,34 +15,99 @@
  */
 
 import * as Preact from '../../../src/preact';
-import {ContainWrapper} from '../../../src/preact/component';
+import {ContainWrapper, useValueRef} from '../../../src/preact/component';
+import {Loading} from '../../../src/loading';
+import {ReadyState} from '../../../src/ready-state';
 import {dict} from '../../../src/utils/object';
+import {forwardRef} from '../../../src/preact/compat';
 import {getData} from '../../../src/event-helper';
 import {parseJson} from '../../../src/json';
-import {useLayoutEffect, useRef, useState} from '../../../src/preact';
-import {useLoad} from '../../../src/preact/context';
+import {useAmpContext, useLoading} from '../../../src/preact/context';
+import {
+  useCallback,
+  useEffect,
+  useImperativeHandle,
+  useLayoutEffect,
+  useRef,
+  useState,
+} from '../../../src/preact';
 
 const NO_HEIGHT_STYLE = dict();
 
 /**
- * @param {!InstagramPropsDef} props
+ * @param {!InstagramDef.Props} props
+ * @param {{current: (!InstagramDef.Api|null)}} ref
  * @return {PreactDef.Renderable}
+ * @template T
  */
-export function Instagram({
-  shortcode,
-  captioned,
-  title,
-  requestResize,
-  loading,
-  onLoad,
-  ...rest
-}) {
-  const load = useLoad(loading, /* unlayoutOnPause */ true);
+export function InstagramWithRef(
+  {
+    shortcode,
+    captioned,
+    title,
+    requestResize,
+    loading: loadingProp,
+    onReadyState,
+    ...rest
+  },
+  ref
+) {
+  const {playable} = useAmpContext();
+  const loading = useLoading(loadingProp);
+  const load = loading !== Loading.UNLOAD;
+
+  const loadedRef = useRef(false);
+  // The `onReadyStateRef` is passed via a ref to avoid the changed values
+  // of `onReadyState` re-triggering the side effects.
+  const onReadyStateRef = useValueRef(onReadyState);
+  const setLoaded = useCallback(
+    (value) => {
+      if (value !== loadedRef.current) {
+        loadedRef.current = value;
+        const onReadyState = onReadyStateRef.current;
+        onReadyState?.(value ? ReadyState.COMPLETE : ReadyState.LOADING);
+      }
+    },
+    [onReadyStateRef]
+  );
 
   const iframeRef = useRef(null);
   const [heightStyle, setHeightStyle] = useState(NO_HEIGHT_STYLE);
   const [opacity, setOpacity] = useState(0);
 
+  // Component API: InstagramDef.Api.
+  useImperativeHandle(
+    ref,
+    () => ({
+      // Standard Bento
+      get readyState() {
+        return loadedRef.current ? ReadyState.COMPLETE : ReadyState.LOADING;
+      },
+    }),
+    []
+  );
+
+  // Reset readyState to "loading" when an iframe is unloaded. Has to be
+  // a `useLayoutEffect` to avoid race condition with a future "load" event.
+  // A race condition can happen when a `useEffect` would be executed
+  // after a future "load" is dispatched.
+  useLayoutEffect(() => {
+    if (!load) {
+      setLoaded(false);
+    }
+  }, [load, setLoaded]);
```

Comment:
This loaded code is really difficult for me to grok. I think what's happening is we want to call `onReadyStateChange` when unloaded->loaded and loaded->unloaded. I don't understand why it can't simply be written like:

```js
const lastLoadedRef = useRef(false);
const [loaded, setLoaded] = useState(false);

if (loaded && !load) {
  onReadyState?.(ReadyState.LOADING);
  lastLoadedRef.current = false;
  setLoaded(false);
} else if (loaded !== lastLoadedRef.current) {
  onReadyState?.(loaded ? ReadyState.COMPLETE : ReadyState.LOADING);
  lastLoadedRef.current = loaded;
}

Reply:
I renamed things a bit here to make it simpler. LMK if that helps. I renamed `load` variable to `mount` - it basically decides whether or not to actually render/mount the iframe.

The goal of this code specifically is to reset `loaded` to `false` because if the iframe is no longer in DOM, it's not loaded either. And when it's added back, it will restart the whole chain of states again.

Your code might be close, but (a) we can't send events from the component's render function and we don't want to send them on each render. So an effect is better for this.